use std::unique_ptr for NodeMap resources

mikemorris opened the pull request:
Properly clean up all resources in `NodeMap::release` to allow the `NodeMap`
object to be garbage collected.

Fixes a memory leak that likely started in https://github.com/mapbox/node-mapbox-gl-native/pull/143 when the `NodeFileSource` object was moved inside the `NodeMap` instead of being an external reference.

/cc @lucaswoj @jfirebaugh @miccolis

lucaswoj commented on platform/node/src/node_map.cpp:
```diff
@@ -278,9 +278,8 @@ void NodeMap::renderFinished() {
-            const_cast<mbgl::StillImage *>(img.get())
+            const_cast<mbgl::StillImage *>(img.release())
         ).ToLocalChecked();
-        img.release();
```
I'm curious -- can you explain this change?

mikemorris replied:
Just moved it a line above so it releases directly into the `void*` in [`Nan::NewBuffer`](https://github.com/nodejs/nan/blob/master/doc/buffers.md#api_nan_new_buffer). I'm actually not sure if there's a guarantee that these arguments will be evaluated in the order I expect.

lucaswoj replied:
AFIK, there are no cross-platform guarantees about the order in which arguments are evaluated

jfirebaugh replied:
Yeah, this should be reverted; C++ does not guarantee order of evaluation for function arguments.

mikemorris replied:
Reverted in https://github.com/mapbox/mapbox-gl-native/commit/6d671db02a1425cb11d06cc40facf7944e27e812

lucaswoj commented:
This looks :shipit: once Travis passes!

lucaswoj commented:
@mikemorris @jfirebaugh is this ready to ship?

mikemorris commented:
Hitting a segfault in `npm test` on Node.js v0.12.7
```
# returns an image
Floating point exception (core dumped)
```

mikemorris commented:
Reproducing the segfault locally on OS X under Node.js v0.12.7, backtrace:
```
  * frame #0: 0x0000000103b2d84e mapbox-gl-native.node`std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > >, void*>*> std::__1::__hash_table<std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > >, std::__1::__unordered_map_hasher<mbgl::Resource, std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > >, mbgl::Resource::Hash, true>, std::__1::__unordered_map_equal<mbgl::Resource, std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > >, std::__1::equal_to<mbgl::Resource>, true>, std::__1::allocator<std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > > > >::find<mbgl::Resource>(this=0x0000000101513c80, __k=0x00000001014107b0) + 142 at __hash_table:2019
    frame #1: 0x0000000103b2ae74 mapbox-gl-native.node`node_mbgl::NodeFileSource::notify(mbgl::Resource const&, std::__1::shared_ptr<mbgl::Response const> const&) [inlined] std::__1::unordered_map<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> >, mbgl::Resource::Hash, std::__1::equal_to<mbgl::Resource>, std::__1::allocator<std::__1::pair<mbgl::Resource const, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > > > >::find(this=0x0000000101513c80 size=0, __k=0x00000001014107b0) + 8 at unordered_map:972
    frame #2: 0x0000000103b2ae6c mapbox-gl-native.node`node_mbgl::NodeFileSource::notify(this=0x0000000101513c70, resource=0x00000001014107b0, response=std::__1::shared_ptr<const mbgl::Response>::element_type @ 0x0000000101516708 strong=2 weak=1) + 28 at node_file_source.cpp:117
    frame #3: 0x0000000103b31945 mapbox-gl-native.node`node_mbgl::NodeRequest::Respond(info=0x00007fff5fbf6538) + 1221 at node_request.cpp:80
    frame #4: 0x0000000103b2b163 mapbox-gl-native.node`Nan::imp::FunctionCallbackWrapper(info=0x00007fff5fbf6578) + 131 at nan_callbacks_12_inl.h:174
    frame #5: 0x000000010013f2ef node`v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&)) + 159
    frame #6: 0x000000010016309b node`v8::internal::Builtin_HandleApiCall(int, v8::internal::Object**, v8::internal::Isolate*) + 507
    frame #7: 0x00003ef786c060a2
    frame #8: 0x00003ef786d35f30
    frame #9: 0x00003ef786c1e8d5
    frame #10: 0x00003ef786d35d57
    frame #11: 0x00003ef786c1e8d5
    frame #12: 0x00003ef786c24ac2
    frame #13: 0x00003ef786d35c3d
    frame #14: 0x00003ef786c1e8d5
    frame #15: 0x00003ef786c1ef3c
    frame #16: 0x00003ef786c1de90
    frame #17: 0x0000000100212eaf node`v8::internal::Invoke(bool, v8::internal::Handle<v8::internal::JSFunction>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*) + 431
    frame #18: 0x0000000100129b9a node`v8::Function::Call(v8::Handle<v8::Value>, int, v8::Handle<v8::Value>*) + 202
    frame #19: 0x0000000100504fb7 node`node::AsyncWrap::MakeCallback(v8::Handle<v8::Function>, int, v8::Handle<v8::Value>*) + 521
    frame #20: 0x00000001005263c5 node`node::After(uv_fs_s*) + 689
    frame #21: 0x00000001006191db node`uv__work_done + 175
    frame #22: 0x000000010061abdc node`uv__async_event + 65
    frame #23: 0x000000010061ad5f node`uv__async_io + 136
    frame #24: 0x0000000100628cd8 node`uv__io_poll + 1526
    frame #25: 0x000000010061b220 node`uv_run + 339
    frame #26: 0x0000000100512ed6 node`node::Start(int, char**) + 339
    frame #27: 0x0000000100001034 node`start + 52
```

mikemorris commented:
I think this may be hitting https://github.com/mapbox/api-gl/issues/260

mikemorris commented:
I'm hitting segfaults like the one below locally pretty consistently by running `lldb platform/node/test/js/map.test.js` on this branch.
```
# returns an image
NodeRequest::Respond (Error)
NodeFileSource::request
libc++abi.dylib: terminating with uncaught exception of type std::__1::system_
error: mutex lock failed: Invalid argument
```
There's something going on with `cancel` calls in `NodeFileSource`, such that I can't figure out where `NodeFileSource::cancel` is actually called, `processCancel` is never being called, and I think the segfault above is from uncanceled requests trying to notify an already-destructed `NodeFileSource`.

I'm having trouble making sense of the `NodeFileSource` code after a few revisions of the `FileSource` code - it seems that `NodeRequest` no longer wraps an `mbgl::Request` object (where does `NodeFileSource::request` get called?) and the cancelation flow was complicated by https://github.com/mapbox/mapbox-gl-native/commit/71921230f06ad559177a66de88a72f7de26d77f2#commitcomment-13909494

Bringing back `observers.erase(observersIt);` in `NodeFileSource::notify` causes a hang FWIW. Think you would be able to lend a hand to help me untangle this @tmpsantos or @kkaefer?

tmpsantos commented:
> Bringing back observers.erase(observersIt); in NodeFileSource::notify causes a hang FWIW. Think you would be able to lend a hand to help me untangle this @tmpsantos or @kkaefer?

Seems to me that `NodeRequest` is outliving the `NodeFileSource`. You are canceling the `mbgl::Request` but the `NodeRequest` living in the node Javascript context is still holding a reference to `NodeFileSource` for calling `source->notify()`.

You can probably make the `NodeFileSource` subscribe to the requests as an observer, so when you destroy the `NodeFileSource` it would unref the requests and unsubscribe as an observer.

When the request notifies in this case, it would be a no-op.

mikemorris commented:
Still a couple TODOs, but the refactored `NodeFileSource` compiles now - having trouble with the `FileSource` thread though, the `v8::Local<v8::Object>` passed into the `NodeFileSource::Impl` constructor seems to get garbage collected by the time it arrives. Possible I could work around this by keeping it in the outer shell, will experiment more tomorrow.
```
Process 56986 stopped
* thread #8: tid = 0x397797, 0x0000000100117863 node`v8::V8::GlobalizeReference(v8::internal::Object**) + 35, name = 'FileSource', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x0000000100117863 node`v8::V8::GlobalizeReference(v8::internal::Object**) + 35
node`v8::V8::GlobalizeReference:
->  0x100117863 <+35>: cmpl   $0x1, (%rbx)
    0x100117866 <+38>: je     0x1001178b8               ; <+120>
    0x100117868 <+40>: leaq   0x5d8104(%rip), %rdi      ; v8::internal::V8::has_fatal_error_
    0x10011786f <+47>: leaq   0x5d80fc(%rip), %rsi      ; v8::internal::V8::has_been_disposed_
(lldb) bt
* thread #8: tid = 0x397797, 0x0000000100117863 node`v8::V8::GlobalizeReference(v8::internal::Object**) + 35, name = 'FileSource', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x0000000100117863 node`v8::V8::GlobalizeReference(v8::internal::Object**) + 35
    frame #1: 0x0000000101ac9718 mapbox-gl-native.node`void mbgl::util::Thread<mbgl::NodeFileSource::Impl>::run<std::__1::tuple<v8::Local<v8::Object>&>, 0ul>(mbgl::util::ThreadContext, std::__1::tuple<v8::Local<v8::Object>&>&&, std::__1::integer_sequence<unsigned long, 0ul>) [inlined] v8::Persistent<v8::Object>::New(v8::Handle<v8::Object>) + 184 at v8.h:4228
    frame #2: 0x0000000101ac9710 mapbox-gl-native.node`void mbgl::util::Thread<mbgl::NodeFileSource::Impl>::run<std::__1::tuple<v8::Local<v8::Object>&>, 0ul>(mbgl::util::ThreadContext, std::__1::tuple<v8::Local<v8::Object>&>&&, std::__1::integer_sequence<unsigned long, 0ul>) [inlined] void Nan::PersistentBase<v8::Object>::Reset<v8::Object>(v8::Local<v8::Object> const&) + 5 at nan_persistent_pre_12_inl.h:43
    frame #3: 0x0000000101ac970b mapbox-gl-native.node`void mbgl::util::Thread<mbgl::NodeFileSource::Impl>::run<std::__1::tuple<v8::Local<v8::Object>&>, 0ul>(mbgl::util::ThreadContext, std::__1::tuple<v8::Local<v8::Object>&>&&, std::__1::integer_sequence<unsigned long, 0ul>) [inlined] mbgl::NodeFileSource::Impl::Impl(v8::Local<v8::Object>) + 64 at node_file_source.cpp:44
    frame #4: 0x0000000101ac96cb mapbox-gl-native.node`void mbgl::util::Thread<mbgl::NodeFileSource::Impl>::run<std::__1::tuple<v8::Local<v8::Object>&>, 0ul>(mbgl::util::ThreadContext, std::__1::tuple<v8::Local<v8::Object>&>&&, std::__1::integer_sequence<unsigned long, 0ul>) [inlined] mbgl::NodeFileSource::Impl::Impl(options_=<unavailable>) at node_file_source.cpp:43
    frame #5: 0x0000000101ac96cb mapbox-gl-native.node`void mbgl::util::Thread<mbgl::NodeFileSource::Impl>::run<std::__1::tuple<v8::Local<v8::Object>&>, 0ul>(this=0x0000000100c01780, context=<unavailable>, params=<unavailable>, (null)=<unavailable>) + 107 at thread.hpp:123
    frame #6: 0x0000000101ac960b mapbox-gl-native.node`std::__1::__thread_proxy<std::__1::tuple<mbgl::util::Thread<mbgl::NodeFileSource::Impl>::Thread<v8::Local<v8::Object>&>(mbgl::util::ThreadContext const&, v8::Local<v8::Object>&&&)::'lambda'()> >(void*, void*) [inlined] mbgl::util::Thread<mbgl::NodeFileSource::Impl>::Thread<v8::Local<v8::Object>&>(mbgl::util::ThreadContext const&, v8::Local<v8::Object>&&&)::'lambda'()::operator()() const + 94 at thread.hpp:106
    frame #7: 0x0000000101ac95ad mapbox-gl-native.node`std::__1::__thread_proxy<std::__1::tuple<mbgl::util::Thread<mbgl::NodeFileSource::Impl>::Thread<v8::Local<v8::Object>&>(mbgl::util::ThreadContext const&, v8::Local<v8::Object>&&&)::'lambda'()> >(void*, void*) [inlined] std::__1::__invoke<mbgl::util::Thread<mbgl::NodeFileSource::Impl>::Thread<v8::Local<v8::Object>&>(mbgl::util::ThreadContext const&, v8::Local<v8::Object>&&&)::'lambda'()>(decltype(std::__1::forward<mbgl::util::Thread<mbgl::NodeFileSource::Impl>::Thread<v8::Local<v8::Object>&>(mbgl::util::ThreadContext const&, v8::Local<v8::Object>&&&)::'lambda'()>(fp)(std::__1::forward<>(fp0))), mbgl::util::Thread<mbgl::NodeFileSource::Impl>::Thread<v8::Local<v8::Object>&>(mbgl::util::ThreadContext const&, v8::Local<v8::Object>&&&)::'lambda'()&&) + 3 at __functional_base:415
    frame #8: 0x0000000101ac95aa mapbox-gl-native.node`std::__1::__thread_proxy<std::__1::tuple<mbgl::util::Thread<mbgl::NodeFileSource::Impl>::Thread<v8::Local<v8::Object>&>(mbgl::util::ThreadContext const&, v8::Local<v8::Object>&&&)::'lambda'()> >(void*, void*) [inlined] _ZNSt3__116__thread_executeIZN4mbgl4util6ThreadINS1_14NodeFileSource4ImplEEC1IJRN2v85LocalINS8_6ObjectEEEEEERKNS2_13ThreadContextEDpOT_EUlvE_JEJEEEvRNS_5tupleIJT_DpT0_EEENS_15__tuple_indicesIJXspT1_EEEE at thread:337
    frame #9: 0x0000000101ac95aa mapbox-gl-native.node`std::__1::__thread_proxy<std::__1::tuple<mbgl::util::Thread<mbgl::NodeFileSource::Impl>::Thread<v8::Local<v8::Object>&>(mbgl::util::ThreadContext const&, v8::Local<v8::Object>&&&)::'lambda'()> >(__vp=0x0000000100c12b20) + 90 at thread:347
    frame #10: 0x00007fff97b2b9b1 libsystem_pthread.dylib`_pthread_body + 131
    frame #11: 0x00007fff97b2b92e libsystem_pthread.dylib`_pthread_start + 168
    frame #12: 0x00007fff97b29385 libsystem_pthread.dylib`thread_start + 13
```

mikemorris commented:
Attempting to use `Nan::Global` to work around the different `v8::Isolate` in the `FileSource` thread, but hitting an issue with the `std::tuple` argument forwarding.
```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/tuple:277:5: error: the parameter for this
      explicitly-defaulted copy constructor is const, but a member or base requires it to be non-const
    __tuple_leaf(const __tuple_leaf& __t) = default;
    ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/tuple:398:14: note: in instantiation of template
      class 'std::__1::__tuple_leaf<0, Nan::Global<v8::Object>, false>' requested here
    : public __tuple_leaf<_Indx, _Tp>...
             ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/tuple:504:10: note: in instantiation of template
      class 'std::__1::__tuple_impl<std::__1::__tuple_indices<0>, Nan::Global<v8::Object> >' requested here
    base base_;
         ^
../src/mbgl/util/thread.hpp:93:25: note: in instantiation of template class 'std::__1::tuple<Nan::Global<v8::Object> >' requested here
    std::tuple<Args...> params = std::forward_as_tuple(::std::forward<Args>(args)...);
                        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3075:32: note: in instantiation of function
      template specialization 'mbgl::util::Thread<mbgl::NodeFileSource::Impl>::Thread<Nan::Global<v8::Object> >' requested here
    return unique_ptr<_Tp>(new _Tp(_VSTD::forward<_Args>(__args)...));
                               ^
../platform/node/src/node_file_source.cpp:14:17: note: in instantiation of function template specialization
      'std::__1::make_unique<mbgl::util::Thread<mbgl::NodeFileSource::Impl>, mbgl::util::ThreadContext, Nan::Global<v8::Object> >' requested here
    thread(std::make_unique<util::Thread<Impl>>(
                ^
1 error generated.
```

mikemorris commented:
After hitting trouble with an `implicit instantiation of undefined template` error trying to reimplement the `Queue`, I had a bit more success dropping all the queues, locks and threads instead, but I'm still hitting a v8 issue related to the `HandleScope`. (Is something happening on a thread that doesn't have a `HandleScope` perhaps? The queue was previously doing some things in `uv_default_loop`.)
```
ok 22 should throw
# returns an error
Process 6840 stopped
* thread #8: tid = 0x3cc1d9, 0x0000000100117bb4 node`v8::HandleScope::HandleScope() + 52, name = 'Map', stop reason = EXC_BAD_ACCESS (code=1, address=0x2de8)
    frame #0: 0x0000000100117bb4 node`v8::HandleScope::HandleScope() + 52
node`v8::HandleScope::HandleScope:
->  0x100117bb4 <+52>: movq   0x2de8(%rbx), %r13
    0x100117bbb <+59>: callq  0x100244e40               ; v8::internal::ThreadId::GetCurrentThreadId()
    0x100117bc0 <+64>: cmpl   %eax, 0x8(%r13)
    0x100117bc4 <+68>: je     0x100117c01               ; <+129>
```

mikemorris commented:
Pushed up an attempt at reimplementing the queue to pop creation of v8 objects back out to the Node.js thread, error spew from implicit instantiation of undefined template looks like this:
```
  CXX(target) Release/obj.target/mapbox-gl-native/platform/node/src/node_file_source.o
../platform/node/src/node_file_source.cpp:28:16: error: implicit instantiation of undefined template
      'mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action>'
    impl->queue->send(Impl::Action { Impl::Action::Add, req });
               ^
../platform/node/src/node_file_source_impl.hpp:10:46: note: template is declared here
namespace util { template <typename T> class AsyncQueue; }
                                             ^
../platform/node/src/node_file_source.cpp:35:16: error: implicit instantiation of undefined template
      'mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action>'
    impl->queue->send(Impl::Action { Impl::Action::Cancel, req });
               ^
../platform/node/src/node_file_source_impl.hpp:10:46: note: template is declared here
namespace util { template <typename T> class AsyncQueue; }
                                             ^
../platform/node/src/node_file_source.cpp:56:10: error: implicit instantiation of undefined template
      'mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action>'
    queue->unref();
         ^
../platform/node/src/node_file_source_impl.hpp:10:46: note: template is declared here
namespace util { template <typename T> class AsyncQueue; }
                                             ^
../platform/node/src/node_file_source.cpp:60:10: error: implicit instantiation of undefined template
      'mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action>'
    queue->stop();
         ^
../platform/node/src/node_file_source_impl.hpp:10:46: note: template is declared here
namespace util { template <typename T> class AsyncQueue; }
                                             ^
../platform/node/src/node_file_source.cpp:68:14: error: implicit instantiation of undefined template
      'mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action>'
        queue->ref();
             ^
../platform/node/src/node_file_source_impl.hpp:10:46: note: template is declared here
namespace util { template <typename T> class AsyncQueue; }
                                             ^
../platform/node/src/node_file_source.cpp:136:20: error: implicit instantiation of undefined template
      'mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action>'
              queue->unref();
                   ^
../platform/node/src/node_file_source_impl.hpp:10:46: note: template is declared here
namespace util { template <typename T> class AsyncQueue; }
                                             ^
../platform/node/src/node_file_source.cpp:136:20: error: implicit instantiation of undefined template
      'mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action>'
              queue->unref();
                   ^
../platform/node/src/node_file_source_impl.hpp:10:46: note: template is declared here
namespace util { template <typename T> class AsyncQueue; }
                                             ^
In file included from ../platform/node/src/node_file_source.cpp:1:
In file included from ../platform/node/src/node_file_source_impl.hpp:3:
In file included from ../platform/node/src/node_file_source.hpp:3:
In file included from ../include/mbgl/storage/file_source.hpp:4:
In file included from ../include/mbgl/storage/response.hpp:4:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/string:439:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:628:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:3075:32: error: implicit instantiation of
      undefined template 'mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action>'
    return unique_ptr<_Tp>(new _Tp(_VSTD::forward<_Args>(__args)...));
                               ^
../platform/node/src/node_file_source.cpp:47:16: note: in instantiation of function template specialization
      'std::__1::make_unique<mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action>, uv_loop_s *, (lambda at
      ../platform/node/src/node_file_source.cpp:47:54)>' requested here
    queue(std::make_unique<Queue>(uv_default_loop(), [this](Action &action) {
               ^
../platform/node/src/node_file_source_impl.hpp:10:46: note: template is declared here
namespace util { template <typename T> class AsyncQueue; }
                                             ^
In file included from ../platform/node/src/node_file_source.cpp:1:
In file included from ../platform/node/src/node_file_source_impl.hpp:3:
In file included from ../platform/node/src/node_file_source.hpp:3:
In file included from ../include/mbgl/storage/file_source.hpp:4:
In file included from ../include/mbgl/storage/response.hpp:4:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/string:439:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:628:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:2457:27: error: implicit instantiation of
      undefined template 'mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action>'
            static_assert(sizeof(_Tp) > 0, "default_delete can not delete incomplete type");
                          ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:2658:13: note: in instantiation of member
      function 'std::__1::default_delete<mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action> >::operator()' requested here
            __ptr_.second()(__tmp);
            ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:2626:46: note: in instantiation of member
      function 'std::__1::unique_ptr<mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action>,
      std::__1::default_delete<mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action> > >::reset' requested here
    _LIBCPP_INLINE_VISIBILITY ~unique_ptr() {reset();}
                                             ^
../platform/node/src/node_file_source.cpp:47:11: note: in instantiation of member function
      'std::__1::unique_ptr<mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action>,
      std::__1::default_delete<mbgl::util::AsyncQueue<mbgl::NodeFileSource::Impl::Action> > >::~unique_ptr' requested here
    queue(std::make_unique<Queue>(uv_default_loop(), [this](Action &action) {
          ^
../platform/node/src/node_file_source_impl.hpp:10:46: note: template is declared here
namespace util { template <typename T> class AsyncQueue; }
                                             ^
8 errors generated.
```

mikemorris commented:
Successfully pushing requests into the from the `Map` through the `FileSource` now, just need to actually complete the request instead of printing the URL!

mikemorris commented:
Reverted 90% of the changes I had previously made, I pretty much went on a journey through all the old bugs we fixed to end up with the current `NodeFileSource`. A few fixes remain, gonna pre-tag so we can bench this.

mikemorris commented:
So now the remaining rare segault is from `RequestHolder` kicking off a `NodeFileSource::cancel` request, which triggers `NodeFileSource::processCancel` on the Node.js thread, which attempts to access `std::unordered_map pending` after it has been deleted by `fs.reset()` in `NodeMap::release`.

Either stopping the queue earlier, locking somewhere, or cleaning up everything in the destructor seem like possible solutions here.

```
* thread #1: tid = 0x46a811, 0x000000010380700e mapbox-gl-native.node`std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > >, void*>*> std::__1::__hash_table<std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > >, std::__1::__unordered_map_hasher<mbgl::Resource, std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > >, mbgl::Resource::Hash, true>, std::__1::__unordered_map_equal<mbgl::Resource, std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > >, std::__1::equal_to<mbgl::Resource>, true>, std::__1::allocator<std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > > > >::find<mbgl::Resource>(this=0x000000010170a368, __k=0x000000010170c850) + 142 at __hash_table:2019, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x6bd801458)
  * frame #0: 0x000000010380700e mapbox-gl-native.node`std::__1::__hash_iterator<std::__1::__hash_node<std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > >, void*>*> std::__1::__hash_table<std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > >, std::__1::__unordered_map_hasher<mbgl::Resource, std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > >, mbgl::Resource::Hash, true>, std::__1::__unordered_map_equal<mbgl::Resource, std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > >, std::__1::equal_to<mbgl::Resource>, true>, std::__1::allocator<std::__1::__hash_value_type<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > > > >::find<mbgl::Resource>(this=0x000000010170a368, __k=0x000000010170c850) + 142 at __hash_table:2019
    frame #1: 0x0000000103804504 mapbox-gl-native.node`mbgl::NodeFileSource::notify(mbgl::Resource const&, std::__1::shared_ptr<mbgl::Response const> const&) [inlined] std::__1::unordered_map<mbgl::Resource, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> >, mbgl::Resource::Hash, std::__1::equal_to<mbgl::Resource>, std::__1::allocator<std::__1::pair<mbgl::Resource const, Nan::Persistent<v8::Object, v8::NonCopyablePersistentTraits<v8::Object> > > > >::find(this=0x000000010170a368 size=67282432, __k=0x000000010170c850) + 8 at unordered_map:972
    frame #2: 0x00000001038044fc mapbox-gl-native.node`mbgl::NodeFileSource::notify(this=0x000000010170a350, resource=0x000000010170c850, response=std::__1::shared_ptr<const mbgl::Response>::element_type @ 0x0000000102002e58 strong=2 weak=1) + 28 at node_file_source.cpp:123
    frame #3: 0x000000010380ba6f mapbox-gl-native.node`mbgl::NodeRequest::Respond(info=0x00007fff5fbf6528) + 1311 at node_request.cpp:79
    frame #4: 0x00000001038047e3 mapbox-gl-native.node`Nan::imp::FunctionCallbackWrapper(info=0x00007fff5fbf6568) + 131 at nan_callbacks_12_inl.h:174
    frame #5: 0x0000000100160a0f node`v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&)) + 159
    frame #6: 0x0000000100188fa4 node`v8::internal::MaybeHandle<v8::internal::Object> v8::internal::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::internal::(anonymous namespace)::BuiltinArguments<(v8::internal::BuiltinExtraArguments)1>&) + 1060
    frame #7: 0x000000010018bc0d node`v8::internal::Builtin_HandleApiCall(int, v8::internal::Object**, v8::internal::Isolate*) + 61
    frame #8: 0x000029aca9d060bb
    frame #9: 0x000029aca9e516e2
    frame #10: 0x000029aca9d19494
    frame #11: 0x000029aca9e5148f
    frame #12: 0x000029aca9d19494
    frame #13: 0x000029aca9d19f79
    frame #14: 0x000029aca9d189e2
    frame #15: 0x00000001002bb688 node`v8::internal::Invoke(bool, v8::internal::Handle<v8::internal::JSFunction>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*) + 728
    frame #16: 0x0000000100147904 node`v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*) + 276
    frame #17: 0x00000001006360dd node`node::AsyncWrap::MakeCallback(v8::Local<v8::Function>, int, v8::Local<v8::Value>*) + 547
    frame #18: 0x0000000100662065 node`node::After(uv_fs_s*) + 578
    frame #19: 0x0000000100768f02 node`uv__work_done + 175
    frame #20: 0x000000010076a9f4 node`uv__async_event + 65
    frame #21: 0x000000010076ab77 node`uv__async_io + 136
    frame #22: 0x0000000100778f96 node`uv__io_poll + 1554
    frame #23: 0x000000010076b02b node`uv_run + 321
    frame #24: 0x000000010064825c node`node::Start(int, char**) + 583
    frame #25: 0x0000000100001834 node`start + 52
```

tmpsantos commented:
> Either stopping the queue earlier, checking for existence of resources in NodeRequest::Respond, locking somewhere, or cleaning up everything in the destructor seem like possible solutions here.

The issue here seems to me that `NodeRequest` is calling a `NodeFileSource` that was already destroyed. I would just walk the list of `NodeRequest`s at `~NodeFileSource` and call a `request->invalidate()` that would clear this reference.

No locking is needed because both live in the same thread (main).

mikemorris commented:
Seeing another segfault related to destruction of the `Map` object triggering `uv_async_send` to what I think is a closed/deleted loop - could `map.release()` be sufficient here or would the c++ `Map` object never be cleaned up by the GC in that case?
```
# requires a map style as first argument
ok 11 should throw
Process 9568 stopped
* thread #1: tid = 0x49013b, 0x000000010076aa0b node`uv_async_send + 4, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x78)
    frame #0: 0x000000010076aa0b node`uv_async_send + 4
node`uv_async_send:
->  0x10076aa0b <+4>:  cmpl   $0x0, 0x78(%rdi)
    0x10076aa0f <+8>:  jne    0x10076aa31               ; <+42>
    0x10076aa11 <+10>: movl   $0x1, %ecx
    0x10076aa16 <+15>: xorl   %eax, %eax
(lldb) bt
* thread #1: tid = 0x49013b, 0x000000010076aa0b node`uv_async_send + 4, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x78)
  * frame #0: 0x000000010076aa0b node`uv_async_send + 4
    frame #1: 0x00000001040eccfd mapbox-gl-native.node`mbgl::util::RunLoop::stop() [inlined] uv::async::send() + 9 at uv_detail.hpp:124
    frame #2: 0x00000001040eccf4 mapbox-gl-native.node`mbgl::util::RunLoop::stop() [inlined] void mbgl::util::RunLoop::invoke<mbgl::util::RunLoop::stop()::$_1
>(mbgl::util::RunLoop::stop()::$_1&&) + 190 at run_loop.hpp:41
    frame #3: 0x00000001040ecc36 mapbox-gl-native.node`mbgl::util::RunLoop::stop(this=<unavailable>) + 38 at run_loop.cpp:33
    frame #4: 0x0000000104069776 mapbox-gl-native.node`mbgl::util::Thread<mbgl::MapContext>::~Thread(this=0x0000000102c00eb0) + 22 at thread.hpp:143
    frame #5: 0x0000000104065913 mapbox-gl-native.node`mbgl::Map::~Map() [inlined] mbgl::util::Thread<mbgl::MapContext>::~Thread(this=0x0000000102c00eb0) + 13
1 at thread.hpp:142
    frame #6: 0x000000010406590b mapbox-gl-native.node`mbgl::Map::~Map() [inlined] std::__1::default_delete<mbgl::util::Thread<mbgl::MapContext> >::operator(_
_ptr=0x0000000102c00eb0)(mbgl::util::Thread<mbgl::MapContext>*) const at memory:2459
    frame #7: 0x000000010406590b mapbox-gl-native.node`mbgl::Map::~Map() [inlined] std::__1::unique_ptr<mbgl::util::Thread<mbgl::MapContext>, std::__1::defaul
t_delete<mbgl::util::Thread<mbgl::MapContext> > >::reset(mbgl::util::Thread<mbgl::MapContext>*) + 17 at memory:2658
    frame #8: 0x00000001040658fa mapbox-gl-native.node`mbgl::Map::~Map() [inlined] std::__1::unique_ptr<mbgl::util::Thread<mbgl::MapContext>, std::__1::defaul
t_delete<mbgl::util::Thread<mbgl::MapContext> > >::~unique_ptr() at memory:2626
    frame #9: 0x00000001040658fa mapbox-gl-native.node`mbgl::Map::~Map() [inlined] std::__1::unique_ptr<mbgl::util::Thread<mbgl::MapContext>, std::__1::defaul
t_delete<mbgl::util::Thread<mbgl::MapContext> > >::~unique_ptr() at memory:2626
    frame #10: 0x00000001040658fa mapbox-gl-native.node`mbgl::Map::~Map(this=0x0000000102c019e0) + 106 at map.cpp:30
    frame #11: 0x000000010400a0ed mapbox-gl-native.node`mbgl::NodeMap::release() [inlined] std::__1::default_delete<mbgl::Map>::operator(__ptr=0x0000000102c01
9e0)(mbgl::Map*) const + 8 at memory:2459
    frame #12: 0x000000010400a0e5 mapbox-gl-native.node`mbgl::NodeMap::release() [inlined] std::__1::unique_ptr<mbgl::Map, std::__1::default_delete<mbgl::Map>
 >::reset(mbgl::Map*) + 17 at memory:2658
    frame #13: 0x000000010400a0d4 mapbox-gl-native.node`mbgl::NodeMap::release(this=0x0000000102c01930) + 36 at node_map.cpp:411
    frame #14: 0x0000000104008afa mapbox-gl-native.node`mbgl::NodeMap::Release(info=0x00007fff5fbfe8e8) + 58 at node_map.cpp:394
    frame #15: 0x000000010400a853 mapbox-gl-native.node`Nan::imp::FunctionCallbackWrapper(info=0x00007fff5fbfe928) + 131 at nan_callbacks_12_inl.h:174
    frame #16: 0x0000000100160a0f node`v8::internal::FunctionCallbackArguments::Call(void (*)(v8::FunctionCallbackInfo<v8::Value> const&)) + 159
    frame #17: 0x0000000100188fa4 node`v8::internal::MaybeHandle<v8::internal::Object> v8::internal::HandleApiCallHelper<false>(v8::internal::Isolate*, v8::in
ternal::(anonymous namespace)::BuiltinArguments<(v8::internal::BuiltinExtraArguments)1>&) + 1060
    frame #18: 0x000000010018bc0d node`v8::internal::Builtin_HandleApiCall(int, v8::internal::Object**, v8::internal::Isolate*) + 61
    frame #19: 0x00000b21ba5060bb
    frame #20: 0x00000b21ba643edd
    frame #21: 0x00000b21ba51f5c7
    frame #22: 0x00000b21ba5ea831
    frame #23: 0x00000b21ba519494
    frame #24: 0x00000b21ba60ff9b
    frame #25: 0x00000b21ba51f5c7
    frame #26: 0x00000b21ba5ea831
    frame #27: 0x00000b21ba629687
    frame #28: 0x00000b21ba519494
    frame #29: 0x00000b21ba51f5c3
    frame #30: 0x00000b21ba5ea831
    frame #31: 0x00000b21ba629408
    frame #32: 0x00000b21ba5ece8a
    frame #33: 0x00000b21ba519f7d
    frame #34: 0x00000b21ba5189e2
    frame #35: 0x00000001002bb688 node`v8::internal::Invoke(bool, v8::internal::Handle<v8::internal::JSFunction>, v8::internal::Handle<v8::internal::Object>,
int, v8::internal::Handle<v8::internal::Object>*) + 728
    frame #36: 0x0000000100147904 node`v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*) + 276
    frame #37: 0x00000001006410f2 node`node::MakeCallback(node::Environment*, v8::Local<v8::Value>, v8::Local<v8::Function>, int, v8::Local<v8::Value>*) + 637
    frame #38: 0x0000000100648457 node`node::CheckImmediate(uv_check_s*) + 98
    frame #39: 0x000000010076f2c3 node`uv__run_check + 33
    frame #40: 0x000000010076b033 node`uv_run + 329
    frame #41: 0x000000010064825c node`node::Start(int, char**) + 583
    frame #42: 0x0000000100001834 node`start + 52
```

mikemorris commented:
@tmpsantos Is it possible that some async calls are happening that expect the existence of a `FileSource` or `View` when they complete? I was destroying the `Map` in the `unique_ptr` before, but the destruction of the other two is new, and we're seeing a similar-looking segfault in a completely unrelated spot now:
```
   /build/gdb-HnfxP_/gdb-7.10/gdb/cp-support.c:1595: demangler-warning: unable to demangle '_ZSt10make_tupleIJPN4mbgl10TileWorkerESt6vectorINS0_4util3ptrINS0_10StyleLayerEEESaIS7_EEPKSt13unordered_mapISsSt10unique_ptrINS0_6BucketESt14default_deleteISC_EESt4hashISsESt8equal_toISsESaISt4pairIKSsSF_EEENS0_15PlacementConfigERZNS4_7RunLoop4bindIZNSS_18invokeWithCallbackIZNS4_6ThreadINS0_6Worker4ImplEE4bindIMSX_FvS2_S9_SQ_SR_St8functionIFvvEEEEEDaT_EUlDpOT_E_RS12_JS2_RS9_SQ_RSR_EEESB_INS0_11WorkRequestESD_IS1E_EEOS16_OT0_DpOT1_EUlS19_E_EEDaS1H_EUlS19_E_EESt5tupleIJDpNSt17__decay_and_stripIS17_E6__typeEEES19_' (demangler failed with signal 11)`
```

tmpsantos commented:
> @tmpsantos Is it possible that some async calls are happening that expect the existence of a FileSource or View when they complete (or could the un-garbage-collected NodeFileSource have previously been keeping the libuv event loop alive)? I was destroying the Map in a unique_ptr before, but the destruction of the other two is new, and we're seeing a similar-looking segfault in a completely unrelated spot now:

@jfirebaugh found a bug that could be causing these crashes:

https://github.com/mapbox/mapbox-gl-native/commit/747e0f18499607fdf2d2d917d38f9e66d19f1e9f
https://github.com/mapbox/mapbox-gl-native/commit/0a95409495f30fba98dd2efee21472cbb2d97668

Full PR: https://github.com/mapbox/mapbox-gl-native/pull/3033

Can you give these two commits a try?

tmpsantos commented:
There is one more thing actually.

I think the main loop for the `NodeFileSource` should be a singleton:
https://github.com/mapbox/mapbox-gl-native/commit/1f67471be3f93f97ae3fc5fba72435d80b49c445

It wasn't before because all the maps were sharing the same `FileSource` but now it makes sense to make it a singleton (it shouldn't be working actually, but somehow it is).

mikemorris commented:
Rebased onto #3033, pushed up as [`node-unique-ptr`](https://github.com/mapbox/mapbox-gl-native/tree/node-unique-ptr).

@tmpsantos Is there anything else that needs to be changed for the singleton loop?

jfirebaugh commented:
The change to a singleton `RunLoop` was fixing an issue I introduced in the previous commit. It was never an issue in previous node-mbgl releases as they have always used `uv_default_loop()` everywhere.

@mikemorris The namespace and style changes look fine, but I'm skeptical about 812d997e2c2762503cec383df5d6cd5fc73c335c. Are you still able to reproduce a memory leak before this change and not after?

mikemorris commented:
@jfirebaugh Yes, without https://github.com/mapbox/mapbox-gl-native/commit/812d997e2c2762503cec383df5d6cd5fc73c335c garbage collection never destroys `NodeMap` during `npm test` or `npm run test-suite`

mikemorris commented:
In the current state of `master` (post #3033), it appears that simply calling `map.release()` in the render callback is enough to keep a reference in scope and prevent garbage collection. However, dropping `map.release()` from the callback results in the following error, possibly from `NodeFileSource` being deleted before the `mbgl::Map` destructor has a chance to cancel outstanding requests.
```
Process 61176 stopped
* thread #53: tid = 0x56132a, 0x00007fff98d2f0ae libsystem_kernel.dylib`__pthread_kill + 10, name = 'Worker', stop reason = signal SIGABRT
    frame #0: 0x00007fff98d2f0ae libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff98d2f0ae <+10>: jae    0x7fff98d2f0b8            ; <+20>
    0x7fff98d2f0b0 <+12>: movq   %rax, %rdi
    0x7fff98d2f0b3 <+15>: jmp    0x7fff98d2a3ef            ; cerror_nocancel
    0x7fff98d2f0b8 <+20>: retq
(lldb) bt
* thread #53: tid = 0x56132a, 0x00007fff98d2f0ae libsystem_kernel.dylib`__pthread_kill + 10, name = 'Worker', stop reason = signal SIGABRT
  * frame #0: 0x00007fff98d2f0ae libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff97b2e500 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff872ca37b libsystem_c.dylib`abort + 129
    frame #3: 0x00000001004430e7 node`uv__loop_init + 552
    frame #4: 0x000000010043fea6 node`uv_loop_new + 36
    frame #5: 0x00000001033b3e7c mapbox-gl-native.node`void mbgl::util::Thread<mbgl::Worker::Impl>::run<std::__1::tuple<> >(mbgl::util::ThreadContext, std::__1::tuple<>&&, std::__1::integer_sequence<unsigned long>) [inlined] uv::loop::loop() + 5 at uv_detail.hpp:50
    frame #6: 0x00000001033b3e77 mapbox-gl-native.node`void mbgl::util::Thread<mbgl::Worker::Impl>::run<std::__1::tuple<> >(mbgl::util::ThreadContext, std::__1::tuple<>&&, std::__1::integer_sequence<unsigned long>) [inlined] uv::loop::loop() at uv_detail.hpp:48
    frame #7: 0x00000001033b3e77 mapbox-gl-native.node`void mbgl::util::Thread<mbgl::Worker::Impl>::run<std::__1::tuple<> >(this=0x0000000100d77e50, context=ThreadContext @ 0x000070000355bea0, params=<unavailable>, (null)=<unavailable>) + 39 at thread.hpp:115
    frame #8: 0x00000001033b3dfb mapbox-gl-native.node`std::__1::__thread_proxy<std::__1::tuple<mbgl::util::Thread<mbgl::Worker::Impl>::Thread<>(mbgl::util::ThreadContext const&)::'lambda'()> >(void*, void*) [inlined] mbgl::util::Thread<mbgl::Worker::Impl>::Thread<>(mbgl::util::ThreadContext const&)::'lambda'()::operator()() const + 94 at thread.hpp:106
    frame #9: 0x00000001033b3d9d mapbox-gl-native.node`std::__1::__thread_proxy<std::__1::tuple<mbgl::util::Thread<mbgl::Worker::Impl>::Thread<>(mbgl::util::ThreadContext const&)::'lambda'()> >(void*, void*) [inlined] std::__1::__invoke<mbgl::util::Thread<mbgl::Worker::Impl>::Thread<>(mbgl::util::ThreadContext const&)::'lambda'()>(decltype(std::__1::forward<mbgl::util::Thread<mbgl::Worker::Impl>::Thread<>(mbgl::util::ThreadContext const&)::'lambda'()>(fp)(std::__1::forward<>(fp0))), mbgl::util::Thread<mbgl::Worker::Impl>::Thread<>(mbgl::util::ThreadContext const&)::'lambda'()&&) + 3 at __functional_base:415
    frame #10: 0x00000001033b3d9a mapbox-gl-native.node`std::__1::__thread_proxy<std::__1::tuple<mbgl::util::Thread<mbgl::Worker::Impl>::Thread<>(mbgl::util::ThreadContext const&)::'lambda'()> >(void*, void*) [inlined] _ZNSt3__116__thread_executeIZN4mbgl4util6ThreadINS1_6Worker4ImplEEC1IJEEERKNS2_13ThreadContextEDpOT_EUlvE_JEJEEEvRNS_5tupleIJT_DpT0_EEENS_15__tuple_indicesIJXspT1_EEEE at thread:337
    frame #11: 0x00000001033b3d9a mapbox-gl-native.node`std::__1::__thread_proxy<std::__1::tuple<mbgl::util::Thread<mbgl::Worker::Impl>::Thread<>(mbgl::util::ThreadContext const&)::'lambda'()> >(__vp=0x0000000100d77fa0) + 90 at thread:347
    frame #12: 0x00007fff97b2b9b1 libsystem_pthread.dylib`_pthread_body + 131
    frame #13: 0x00007fff97b2b92e libsystem_pthread.dylib`_pthread_start + 168
    frame #14: 0x00007fff97b29385 libsystem_pthread.dylib`thread_start + 13
```
I tried experimenting with `options.SetWeak` in `NodeFileSource`, but it always seems to get garbage collected early, when it's still needed for [`Nan::MakeCallback(Nan::New(options), "request", 2, argv);`](https://github.com/mapbox/mapbox-gl-native/blob/master/platform/node/src/node_file_source.cpp#L32). Not calling `options.Reset()` in `WeakCallbackInfo` seems to "work", but I'm pretty sure this would actually be a leak.

mikemorris commented:
@jfirebaugh Pushed up [`node-gc-segfault`](https://github.com/mapbox/mapbox-gl-native/compare/node-stop-leak...node-gc-segfault) where I'm reproducing this `SIGABRT` with `lldb node platform/node/test/render.test.js`

jfirebaugh commented:
I don't get a crash with node v5, but I recognize the backtrace now, and it is caused by resource exhaustion due to no GC.

Here's my understanding of what's happening: normally, a pattern like the following would be handled just fine by modern JS engines:

```
var foo = new Foo();
foo.callback = function () {
  ...
}
```

There's a reference cycle here: `foo` has a reference to the callback function, the function has a reference to its closure, and that closure has a reference to `foo`. But modern JS engines collect reference cycles just fine.

_The above is the exact pattern we're dealing with._ `Foo` is `Map`, the callback function is the `request` method. So why doesn't the cycle get GC'd? I assume it's because `Map` is storing a reference to the `request` method in a `Persistent`, and that doing so prevents V8 from collecting the cycle as it normally would.

We can't use a weak persistent reference because the persistent reference is _expected_ to be the only reference to the `request` callback. We just don't want that to prevent the normal garbage collection of cycles. Is there some sort of "medium strength" reference in V8 -- stronger than a weak persistent reference in that it doesn't automatically get collected if it's the singular reference, but weaker than a strong persistent reference in that it doesn't prevent cycles from getting collected? I've done quite a bit of searching on this topic and haven't found anything that helps answer this question.

mikemorris commented:
> I assume it's because `Map` is storing a reference to the request method in a `Persistent`
The `request` method is stored in the `Persistent options` in `NodeFileSource`, to which `NodeMap` holds a reference. I've dug into V8 a bit and haven't found anything like a "medium-strength" persistent ref yet.

mikemorris commented:
@jfirebaugh I think I'm most of the way to a workable solution by setting `options` as a weak persistent on `NodeMap`, then passing a pointer to the `NodeFileSource` constructor (so there will be two references until garbage collection time), just need to add a safety check of some sort in `NodeFileSource::request` or clean up outstanding requests in the `NodeFileSource` destructor.

mikemorris commented:
This was fixed in https://github.com/mapbox/mapbox-gl-native/commit/c4c3a19d85991b3c2703e7b6561d8c1a3fed3543
